### PR TITLE
PTX output for problem editor

### PR DESCRIPTION
**Pull request by Alex-Jordan**
This gives the problem editor a `PTX` display mode option, and then you get to see the PTX XML for the rendered problem. There are some things about this I am unsure of.

- simply tossing in `PTX` to the select options, when the other select options appear to have more structure.
- if I am using the html.ep syntax correctly
- I used a `textarea` to avoid having to escape ampersands and less than signs. Is this bad?

### CODE REVIEW

**Review by drgrice1 — COMMENTED** [review]
This looks okay.

It is a bit odd to add the `PTX` displayMode in this way, but I don't see a better way to do it.  One thing that is a bit odd is that the other options can be removed from the list by the course configuration, but this can not.  It certainly doesn't make sense to include this in that setting though, as this is not a display mode that would ever be available for students.

There is another thing that is odd with this.  That is having the submit buttons in the problem.  Answers can not be entered, and so the submit and preview buttons don't make sense.  The show correct answers button might still make sense, but it doesn't seem to work right with this for some reason.  The correct answers are not displayed in the correct answer column, although if you click on a cell in that column the popup is still shown with the evaluated answer.

**Comment by Alex-Jordan** [comment]
> It is a bit odd to add the PTX displayMode in this way,

Maybe in addition to the config variable that offers display modes, there could be another config variable that offers display modes only available to someone with access to instructor tools. For now it would only have `PTX`.

**Comment by drgrice1** [comment]
Yeah, that could be done.  But then should the `PTX` display mode be added in the other instructor use places where display modes are selected?  Or is it better to just hard code it here, and leave it at that?

File: templates/RPCRenderFormats/default.html.ep [hunk 1]
```diff
@@ -41,7 +41,11 @@
 				<%= form_for $FORM_ACTION_URL, id => 'problemMainForm', class => 'problem-main-form',
 					name => 'problemMainForm', method => 'POST', begin %>
 					<div id="problem_body" class="problem-content" <%== $PROBLEM_LANG_AND_DIR %>>
-						%== $problemText
+						% if ($displayMode eq 'PTX') {
+							<textarea readonly style="width:100%;" rows="25"><%= "\n$problemText\n"%></textarea>
```
[the remark this answers is not in the document]
Comment (Alex-Jordan) [reply templates/RPCRenderFormats/default.html.ep:45]: > I don't see a need for the new lines either. Why are those added?

Artifacts from an earlier iteration, trying to resolve an issue that was resolved by other means. I'll remove them.

[the remark this answers is not in the document]
Comment (Alex-Jordan) [reply templates/RPCRenderFormats/default.html.ep:45]: I like it with `pre` now that I understand about the escaping. Although `class="w-100"` has no visible effect.

[the remark this answers is not in the document]
Comment (drgrice1) [reply templates/RPCRenderFormats/default.html.ep:45]: `class="w-100"` does the same thing that `style="width:100%"` does.  However, it seems that `width:100%` doesn't do anything for a `pre` tag.  So you can just drop that.

File: templates/ContentGenerator/Instructor/PGProblemEditor/view_form.html.ep [hunk 2]
```diff
@@ -19,7 +19,7 @@
 			class => 'col-form-label col-auto' =%>
 		<div class="col-auto">
 			<%= select_field 'action.view.displayMode' => [
-				map { [ $_ => $_, $_ eq $c->{displayMode} ? (selected => undef) : () ] } @{ $ce->{pg}{displayModes}}
+				map { [ $_ => $_, $_ eq $c->{displayMode} ? (selected => undef) : () ] } @{ $ce->{pg}{displayModes}}, 'PTX'
```
[the remark this answers is not in the document]
Comment (Alex-Jordan) [reply templates/ContentGenerator/Instructor/PGProblemEditor/view_form.html.ep]: Done.

File: templates/RPCRenderFormats/default.html.ep (hunk 1 above)
[the remark this answers is not in the document]
Comment (Alex-Jordan) [reply templates/RPCRenderFormats/default.html.ep]: I changed it to just a plain simple `pre`.

**Comment by Alex-Jordan** [comment]
I made the changes you suggested. And you are right that I don't want to provide PTX display options anywhere but here, so I left it hard coded.

Also while it would seem like showing correct answers should be possible and would be useful, I am content without any of the submit buttons being there. I don't plan to work on that and I just took away all three buttons if the display mode is PTX.

**Comment by pstaabp** [comment]
> I could also add 'TeX' to that list if an author just wants a quick peek at the output tex code for a problem without using the hardcopy tab.

This could actually be handy to testing purposes.

**Comment by Alex-Jordan** [comment]
Now there is
```
$pg{additionalPGEditorDisplayModes}
```
with `'PTX'` and `'tex'`, and you can see both using the editor's View tab.

It still hard codes `'PTX'` and `'tex'` for wrapping the output in a `<pre>` and not showing any submission buttons. I am not sure if those should be handled by `$pg{additionalPGEditorDisplayModes}` too.

**Review by drgrice1 — APPROVED** [review]
This looks good to me.

I don't think that the display of correct answers can be fixed in a good easy way for these display modes.  You would have to create a new parameter to pass through the render_rpc endpoint, that would be passed to the WeBWorK::HTML::AttemptsTable constructor to override the display mode back to one of the display modes that it accepts.

An easy, but in my opinion not good, way is to hack in a hard coded case for these in the attempts table that basically uses the MathJax display mode for them.

**Comment by pstaabp** [comment]
When running this on a graph tool problem, I get the following error
```
Failed to generate png file. at [PG]/lib/LaTeXImage.pm line 311.
```
only the first time on View/Reload for PTX mode.  Any other time, I don't see that error.  

Is this a timing problem?  That is, I wonder if the first time the image hasn't been generated yet, so isn't in the tmp directory, throwing an error, but the next time it is. 

And I get the error with a Randomize Seed as well.

**Comment by drgrice1** [comment]
I am not getting that error with any graph tool problems in my testing with this pull request.

**Comment by drgrice1** [comment]
I even delete all of webworks temporary files for the course after loading the problem in the editor so that there wouldn't be any cached images on the server.  Then I reloaded with `PTX` mode, and did not get the error.  I also see that the `.tgz` file is generated for the problem as it should be after doing that.

I tested in the same way with the `tex` display mode, and don't get the error, and see that the `.pdf` file for that is generated.

**Review by pstaabp — APPROVED** [review]

**Pull request merged**